### PR TITLE
Feat: Scheduler

### LINKED ISSUE #75: Delete information after 12 months from consent
Reminder at 11 months

### LINKED ISSUE #108: Backend: Implement handling of account expiration date for citizens

### PR DESCRIPTION
**Pull request by amundlrohne**
Implements schedules with CRON. Structure and initialisation is done. 

Schedules: 
- Notify citizens when there is 1 month remaining of general consent. 
- Delete citizens when today === expirationDate. 
- Notify employees when insight work period is ending, tell them to anonymise external work. 

Still missing schedule for anonymising candidatures, but this will have to wait until insight consents are properly implemented.

### CODE REVIEW

File: apps/api/src/schedulers/index.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import citizenSchedules from './citizenScheduler';
+import insightSchedules from './insightScheduler';
+
+const setupScheduler = () => {
+    citizenSchedules();
+    insightSchedules();
```
Comment (perkynades) [inline apps/api/src/schedulers/index.ts]: Consider renaming these to something like "setupCitizenSchedules", as the current sounds like retriving the current schedule objects

File: apps/api/src/schedulers/citizenScheduler.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    const citizenService = new CitizenService(database);
+    const citizens = await citizenService.getCitizensWithExpirationDate(dateOneMonthFromNow);
+
+    if (!citizens) return;
+
+    const messageService = new MessageService(database);
+    const messages = await messageService.createCitizenExpirationNotification(citizens);
+
+    console.log(messages);
```
Comment (perkynades) [inline apps/api/src/schedulers/citizenScheduler.ts]: Remove log?

File: apps/api/src/schedulers/citizenScheduler.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+const citizenExpirationDateDeletion = cron.schedule('0 0 * * *', async () => {
+    const dateToday = new Date().toISOString().slice(0, 10);
+
+    const citizenService = new CitizenService(database);
+    const citizens = await citizenService.getCitizensWithExpirationDate(dateToday);
+
+    if (!citizens) return;
+
+    await citizenService.deleteMany(citizens);
+    console.log('DELETED CITIZENS');
```
Comment (perkynades) [inline apps/api/src/schedulers/citizenScheduler.ts]: Remove log?

File: apps/api/src/schedulers/insightScheduler.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+const insightExpiredNotifyEmployees = cron.schedule('0 0 * * *', async () => {
+    const dateToday = new Date().toISOString().slice(0, 10);
+
+    const insightService = new InsightService(database);
+    const insights = await insightService.search({
+        where: { end: dateToday },
+        relations: ['project', 'project.members']
+    });
+
+    console.log(insights);
```
Comment (perkynades) [inline apps/api/src/schedulers/insightScheduler.ts]: Remove

File: apps/api/src/schedulers/insightScheduler.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+    const messageService = new MessageService(database);
+    const messages = await messageService.createInsightExpirationNotification(insights);
+
+    console.log(messages);
```
Comment (perkynades) [inline apps/api/src/schedulers/insightScheduler.ts]: Remove

File: apps/api/src/services/InsightService.ts [hunk 6]
```diff
@@ -8,9 +8,10 @@ import BaseService from './BaseService';
 
 export interface IInsightSearch {
     where: {
-        project: string | string[];
+        project?: string | string[];
+        end?: string;
     };
-    relations: string | string[];
+    relations?: string | string[];
```
Comment (perkynades) [inline apps/api/src/services/InsightService.ts:14]: Should all of these be optional?

**Review by perkynades — CHANGES_REQUESTED** [review]
Nice work! Just look at my suggestions and you are good to go!

File: apps/api/src/services/InsightService.ts (hunk 6 above)
Comment (amundlrohne) [reply apps/api/src/services/InsightService.ts:14]: Yes, if not you will be required to include all fields when searching.

File: apps/api/src/schedulers/index.ts (hunk 1 above)
Comment (amundlrohne) [reply apps/api/src/schedulers/index.ts]: Done

File: apps/api/src/schedulers/citizenScheduler.ts (hunk 2 above)
Comment (amundlrohne) [reply apps/api/src/schedulers/citizenScheduler.ts]: Done

File: apps/api/src/schedulers/citizenScheduler.ts (hunk 3 above)
Comment (amundlrohne) [reply apps/api/src/schedulers/citizenScheduler.ts]: Done

File: apps/api/src/schedulers/insightScheduler.ts (hunk 4 above)
Comment (amundlrohne) [reply apps/api/src/schedulers/insightScheduler.ts]: Done

File: apps/api/src/schedulers/insightScheduler.ts (hunk 5 above)
Comment (amundlrohne) [reply apps/api/src/schedulers/insightScheduler.ts]: Done

**Review by perkynades — APPROVED** [review]
LGTM! 👽

**Pull request merged**